File: server/src/plugins/runtime/python/remote_urls.py
import os
import re
from typing import Any

_LOCALHOST = re.compile(r"127\.0\.0\.1|localhost")


def _rewrite_url(url: str, master: str, user: str | None, password: str | None) -> str:
    rewritten = _LOCALHOST.sub(master, url)

    if (
        user
        and rewritten.startswith("rtsp://")
        and "@" not in rewritten[7:].split("/")[0]
    ):
        rewritten = rewritten.replace("rtsp://", f"rtsp://{user}:{password or ''}@", 1)

    return rewritten


def _walk(value: Any, master: str, user: str | None, password: str | None) -> Any:
    if isinstance(value, str):
        return _rewrite_url(value, master, user, password)
    if isinstance(value, list):
        return [_walk(v, master, user, password) for v in value]
    if isinstance(value, dict):
        return {k: _walk(v, master, user, password) for k, v in value.items()}
    return value


def rewrite_source_urls_for_remote(urls: Any) -> Any:
    """No-op unless running remote-hosted with a known master address."""
    master = os.environ.get("CAMERAUI_MASTER_ADDRESS")
    if not os.environ.get("PLUGIN_REMOTE_MODE") or not master:
        return urls

    return _walk(
        urls,
        master,
        os.environ.get("CAMERAUI_RTSP_USERNAME"),
        os.environ.get("CAMERAUI_RTSP_PASSWORD"),
    )

File: server/src/plugins/runtime/node/remoteUrls.ts
const LOCALHOST_PATTERN = /127\.0\.0\.1|localhost/g;

function rewriteUrl(url: string, masterAddress: string, rtspUsername?: string, rtspPassword?: string): string {
  let rewritten = url.replace(LOCALHOST_PATTERN, masterAddress);

  if (rtspUsername && rewritten.startsWith('rtsp://') && !rewritten.slice(7).split('/')[0].includes('@')) {
    rewritten = rewritten.replace('rtsp://', `rtsp://${rtspUsername}:${rtspPassword ?? ''}@`);
  }

  return rewritten;
}

function walk<T>(value: T, masterAddress: string, rtspUsername?: string, rtspPassword?: string): T {
  if (typeof value === 'string') {
    return rewriteUrl(value, masterAddress, rtspUsername, rtspPassword) as T;
  }

  if (Array.isArray(value)) {
    return value.map((entry) => walk(entry, masterAddress, rtspUsername, rtspPassword)) as T;
  }

  if (value && typeof value === 'object') {
    const result: Record<string, unknown> = {};
    for (const [key, entry] of Object.entries(value)) {
      result[key] = walk(entry, masterAddress, rtspUsername, rtspPassword);
    }
    return result as T;
  }

  return value;
}

export function rewriteSourceUrlsForRemote<T>(urls: T): T {
  const masterAddress = process.env.CAMERAUI_MASTER_ADDRESS;
  if (!process.env.PLUGIN_REMOTE_MODE || !masterAddress) {
    return urls;
  }

  return walk(urls, masterAddress, process.env.CAMERAUI_RTSP_USERNAME, process.env.CAMERAUI_RTSP_PASSWORD);
}
